docs/proposals/data-binding.md

```diff
@@ -0,0 +1,169 @@
+# Proposal: Data binding support for Rabbitmq
+
+_Owners_: @shafreenAnfar @aashikam @dilanSachi     
+_Reviewers_: @shafreenAnfar @aashikam  
+_Created_: 2022/05/12  
+_Issues_: [#2818](https://github.com/ballerina-platform/ballerina-standard-library/issues/2818) [#2812](https://github.com/ballerina-platform/ballerina-standard-library/issues/2812) [#2880](https://github.com/ballerina-platform/ballerina-standard-library/issues/2880)
+
+## Summary
+
+Data binding helps to access the incoming and outgoing message data in the user's desired type. Similar to the Ballerina HTTP package, subtypes of `anydata` will be the supported types. This proposal discusses ways to provide data binding on the RabbitMQ client and consumer service side.
+
+## Goals
+
+- Improve user experience by adding data binding support for sending messages with `rabbitmq:Client` and receiving messages with `rabbitmq:Service`.
+
+## Motivation
+
+As of now, the Ballerina `rabbitmq` package does not provide direct data binding for sending and receiving messages. Only `rabbitmq:Message` is the supported data type to send and receive messages which only support `byte[]` as the message content type. Therefore, users have to do data manipulations by themselves. With this new feature, the user experience can be improved by introducing data binding to reduce the burden of developers converting byte data to the desired format as discussed in the next section.
+
+## Description
+
+Currently, when sending a message using `rabbitmq:Client`, user has to convert the message to byte[].
+
+```ballerina
+type Person record {|
+    string name;
+    int age;
+|};
+
+Person person = {
+    age: 10,
+    name: "Harry"
+};
+
+// Publish a message 
+check rabbitmqClient->publishMessage({ content: person.toString().toBytes(), routingKey: "MyQueue" });
+```
+
+When receiving the same message,
+
+```ballerina
+service "MyQueue" on rabbitmqListener {
+    remote function onMessage(rabbitmq:Message message) returns rabbitmq:Error? {
+        string messageContent = check string:fromBytes(message.content);
+        Person person = check value:fromJsonStringWithType(messageContent);
+    }
+}
+```
+
+Receiving the message as a request,
+```ballerina
+service "MyQueue" on rabbitmqListener {
+    remote function onRequest(rabbitmq:Message message) returns anydata|rabbitmq:Error? {
```

Comment:
here we can do `string|rabbitmq:Error? `

Reply:
fixed in https://github.com/ballerina-platform/module-ballerinax-rabbitmq/pull/399/commits/3a0c50ad81de8fcbe2dab219b7971de22ca3be2e